File: tasks/02-fix-the-bug/go/buggy_counter.go
package counter

import (
	"sync"
	"time"
)

var current int64
var mu sync.Mutex

func NextID() int64 {
	// 	it's a race condition problem with goroutine where we face a problem with read, write, modify value by allowing a goroutine do a task that run as a background and make it unsyncronized
	//  then it can make goroutine accessing current value simultaneously not in an order
	// 	it can lead to lost an update of value and return the same ID
	// 	so we need to make it as atomic operation
	// 	using Mutex to make it as a syncronize before we do an operation
	mu.Lock() // lock a code and allow only one goroutine at a time to make it syncronized
	defer mu.Unlock()

	id := current 
	time.Sleep(0)
	current++
	return id
}

File: tasks/02-fix-the-bug/python/buggy_counter.py
# Race condition occurred because multiple threads could read the same _current value
# before any thread incremented it, causing duplicate IDs to be returned.
# Fix by using threading.Lock() to make the read-increment-return sequence atomic, ensuring only one thread can execute the critical section at a time.
# This prevents interleaving of operations that caused the race condition.
import threading
import time

_current = 0
_lock = threading.Lock()


def next_id():
    """Returns a unique ID, incrementing the global counter."""
    global _current
    with _lock:
        value = _current
        time.sleep(0)
        _current += 1
        return value
